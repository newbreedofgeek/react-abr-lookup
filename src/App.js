import React, { Component, lazy, Suspense, StrictMode } from "react";
import { unstable_scheduleCallback } from "scheduler";
import { Spinner } from "./Spinner";
import { SearchForm } from "./SearchForm";

const SearchDetail = lazy(() => import('./SearchDetail'));

const deferredSetState = (self, updater) =>
  unstable_scheduleCallback(() => self.setState(updater));

export default class App extends Component {
  state = { 
    currentId: null, 
    showSearchResult: false
  };

  hdlBackClick = () => {
    this.setState({ currentId: null, showSearchResult: false});
  };

  onResultClick = abn => {
    this.setState({ currentId: abn });
    deferredSetState(this, { showSearchResult: true }); // concurrency (give user click priority)
  }

  componentDidCatch(error) {
    this.setState({ hasError: true });
    console.log(error);
  }

  render() {
    const { hdlBackClick, onResultClick } = this;
    const { showSearchResult, currentId } = this.state;

    if (this.state.hasError) {
      return <h1>Something went wrong (check console)</h1>;
    }

    return (
      <StrictMode>
        {showSearchResult ? (
          <div>
            <button className="onBack" onClick={hdlBackClick}>
              {"ðŸ‘ˆ"} Search Again
            </button>

            <Suspense maxDuration={1200} fallback={<Spinner size="large" />}>
              <SearchDetail id={currentId} />
            </Suspense>
          </div>
        ) : (
          <SearchForm 
            onResultClick={onResultClick}
            loadingId={currentId} />
        )}
      </StrictMode>
    );
  }
}
