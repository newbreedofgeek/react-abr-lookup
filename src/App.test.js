import React from 'react';
import App from './App';
import renderer from 'react-test-renderer';

test('App renders as intended', () => {
  const component = renderer.create(<App />);
  
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});