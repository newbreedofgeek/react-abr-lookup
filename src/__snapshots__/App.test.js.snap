// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders as intended 1`] = `
Array [
  <h1>
    ABR Lookup Service
</h1>,
  <input
    className="searchBox"
    onChange={[Function]}
    onKeyPress={[Function]}
    type="text"
    value=""
/>,
  <button
    className="frmBtn"
    onClick={[Function]}
>
    Search
</button>,
  <button
    className="frmBtn"
    onClick={[Function]}
>
    Reset
</button>,
  <div
    className="schType"
>
    <span>
        Search Type
    </span>
    <ul>
        <li
            className="sel"
            onClick={[Function]}
        >
            name
        </li>
        <li
            className={null}
            onClick={[Function]}
        >
            abn
        </li>
        <li
            className={null}
            onClick={[Function]}
        >
            acn
        </li>
    </ul>
</div>,
  <div
    className="searchResults"
>
    <span>
        Enter a search term and hit search...
    </span>
</div>,
]
`;

exports[`App renders as intended 2`] = `
Array [
  <h1>
    ABR Lookup Service
</h1>,
  <input
    className="searchBox"
    onChange={[Function]}
    onKeyPress={[Function]}
    type="text"
    value="bunnings"
/>,
  <button
    className="frmBtn"
    onClick={[Function]}
>
    Search
</button>,
  <button
    className="frmBtn"
    onClick={[Function]}
>
    Reset
</button>,
  <div
    className="schType"
>
    <span>
        Search Type
    </span>
    <ul>
        <li
            className="sel"
            onClick={[Function]}
        >
            name
        </li>
        <li
            className={null}
            onClick={[Function]}
        >
            abn
        </li>
        <li
            className={null}
            onClick={[Function]}
        >
            acn
        </li>
    </ul>
</div>,
  <div
    className="searchResults"
>
    <div
        className="spinner"
        style={
            Object {
                "fontSize": "35px",
                "lineHeight": "35px",
              }
        }
    >
        ðŸŒ€
    </div>
</div>,
]
`;

exports[`App renders as intended 3`] = `
Array [
  <h1>
    ABR Lookup Service
</h1>,
  <input
    className="searchBox"
    onChange={[Function]}
    onKeyPress={[Function]}
    type="text"
    value="bunnings"
/>,
  <button
    className="frmBtn"
    onClick={[Function]}
>
    Search
</button>,
  <button
    className="frmBtn"
    onClick={[Function]}
>
    Reset
</button>,
  <div
    className="schType"
>
    <span>
        Search Type
    </span>
    <ul>
        <li
            className="sel"
            onClick={[Function]}
        >
            name
        </li>
        <li
            className={null}
            onClick={[Function]}
        >
            abn
        </li>
        <li
            className={null}
            onClick={[Function]}
        >
            acn
        </li>
    </ul>
</div>,
  <div
    className="searchResults"
>
    <div
        className="entity"
        onClick={[Function]}
    >
        <div
            className="main"
        >
            <div
                className="title"
            >
                BUNNINGS
            </div>
            <div
                className="info"
            >
                ABN: 26008672179
            </div>
        </div>
        <div
            className="hover"
        >
            ðŸ‘‰
        </div>
    </div>
    <div
        className="entity"
        onClick={[Function]}
    >
        <div
            className="main"
        >
            <div
                className="title"
            >
                BUNNINGS HANDICRAFTS
            </div>
            <div
                className="info"
            >
                ABN: 63099797977
            </div>
        </div>
        <div
            className="hover"
        >
            ðŸ‘‰
        </div>
    </div>
</div>,
]
`;
